### PR TITLE
Add private key, simple file hosting, S3 hosting and revocation script

### PR DESCRIPTION
**Pull request by 101100**
Thanks for this script! I like how it is very transparent about how it functions, so I wanted to contribute some features.

I wanted to automate the process more while maintaining the ability to function as it does now.  I've added the following capabilities:
- the script can sign the requests itself if it is given the private key
- the script can provide a file and tell the user where to host it (for already running servers)
- the script can upload to an Amazon S3 bucket (for Amazon hosted solutions)

Also, I added a revocation script (since Amazon only allows 2048 bits, so I had to revoke my 4096 bit one).

### CODE REVIEW

**Comment by diafygi** [comment]
Howdy, thanks for the pull request. I feel like this is too much for one pull request.

First, I'm not wild about asking for private keys (even optionally). This script is not audited and I don't think it should ever be trusted with private keys. If you want to give a client access to your private keys, please use the official letsencrypt client (which has had a lot more scrutiny).

Second, not wild about the file-based option. This script is not meant to be run on your server. I'm open to a flag that prints out a command you can copy and paste into your server, but that should be a separate pull request (with minimal code changes).

Third, definitely not wild about the s3-specific option. I want to keep this script as small and readable as possible, and adding vendor-specific options is out of scope.

Fourth, thanks for the revoke_crt.py script! Would be happy to review a separate revoke_crt.py pull request that doesn't have the private-key option. Also, please inline the get_signatures() function since it's not used in a loop.

Anyway, thanks very much for the pull request, but I'm going to close it due to out-of-scope options. Would love a pull request for a simple revoke_crt.py script and a simple file-based flag and printout!

File: revoke_crt.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python
+import argparse, subprocess, json, os, urllib2, sys, base64, binascii, ssl, \
+    hashlib, tempfile, re, time, copy, textwrap, copy
+
+
+def get_signatures(needed_signatures, user_step_number, private_key=None):
```
Comment (diafygi) [inline revoke_crt.py:6]: This is only called once, so should be inlined

**Comment by 101100** [comment]
@diafygi I've made pull requests for the two pieces you'd like to keep.  I'm going to continue working on the version that allows private keys in my own repo.  I'll rename it, give credit and try to explain the philosophical differences so there won't be any confusion.

**Comment by ecc256** [comment]
@diafygi,
The idea is to call "client provided scripts" instead (or in addition to) showing messages like:
"STEP 2: You need to sign some files (replace 'user.key' with your user private key).\n"
Don’t know how to do it right, tho.
May be, if a script "STEP-2" exists, call it?

**Comment by diafygi** [comment]
@101100 Sounds good. Open source at work! FYI, it might be better to start from scratch if you assume that you'll get a private account key. There's a ton of optimizations you can make in the code (reducing printouts, not having to delay making ACME API requests, removing NamedTemporaryFile entirely, etc.), so I wouldn't be surprised if you could get the script under 200 lines of code with a private key. I think that would be better than trying to shim it into this script to have the option of both.

**Comment by kuba** [comment]
Sorry for shameless plug, but you might be interested in https://github.com/kuba/simp_le, @ecc256.

**Comment by diafygi** [comment]
@kuba Cool! Will take a look!

**Comment by ecc256** [comment]
@kuba,
I'm done tailoring @diafygi script for my needs - it does everything now without prompts and works on M$ OS.
I'll take a look at your too, might find something interesting.
My plan was to write plugin for official client, actually.

**Pull request closed without merging**